File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/PIDController.java
package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.util.ElapsedTime;

public class PIDController {
    DcMotor motor;
    double kP, kI, kD;

    ElapsedTime timer;
    int error;
    int lastError;
    double derivative;
    double integralSum;

    int reference;

    public PIDController(double kp, double ki, double kd, DcMotor dcMotor) {
        kP = kp;
        kI = ki;
        kD = kd;
        motor = dcMotor;
        timer = new ElapsedTime();
    }

    public void run() {
        error =  reference - motor.getCurrentPosition();
        derivative = (error - lastError) / timer.seconds();
        integralSum += (error * timer.seconds());

        motor.setPower((kP * error) + (kI * integralSum) + (kD * derivative));

        lastError = error;

        timer.reset();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/PIDController.kt
package org.firstinspires.ftc.teamcode
//
//import com.qualcomm.robotcore.hardware.DcMotor
//import com.qualcomm.robotcore.util.ElapsedTime
//
//class PIDController(val kP: Double, val kI: Double, val kD: Double, val motor: DcMotor) {
//    private val timer = ElapsedTime();
//    private var error = 0
//    private var lastError = 0
//    private var derivative = 0.0
//    private var integralSum = 0.0
//    private var on = false
//    var reference = 0
//
//
//
//    fun run() {
//        error =  reference - motor.currentPosition
//        derivative = (error - lastError) / timer.seconds()
//        integralSum += (error * timer.seconds());
//
//        motor.power = (kP * error) + (kI * integralSum) + (kD * derivative)
//
//        lastError = error
//
//        timer.reset()
//    }
//
//}